library_generation/utilities.sh

```diff
@@ -0,0 +1,176 @@
+}
+
+download_gapic_generator_parent_pom() {
+  gapic_generator_version=$1
+  cd "$working_directory"
+  if [[ "$gapic_generator_version" == *"-SNAPSHOT" ]]; then
+    # get SNAPSHOT from maven local repository.
+    parent_pom=$HOME/.m2/repository/com/google/api/gapic-generator-java-pom-parent/$gapic_generator_version/gapic-generator-java-pom-parent-$gapic_generator_version.pom
+    if [ ! -f "$parent_pom" ]; then
+      echo "Can't copy gapic-generator-java-pom-parent-$gapic_generator_version.pom from maven local repository."
+      exit 1
+    fi
+    cp "$parent_pom" "parent-pom-$gapic_generator_version.xml"
+  fi
+  if [ ! -f "parent-pom-$gapic_generator_version.xml" ]; then
+    curl -LJ -o "parent-pom-$gapic_generator_version.xml" "https://repo1.maven.org/maven2/com/google/api/gapic-generator-java-pom-parent/$gapic_generator_version/gapic-generator-java-pom-parent-$gapic_generator_version.pom"
+  fi
+}
+
+get_grpc_version() {
+  gapic_generator_version=$1
+  # get grpc version from gapic-generator-java-pom-parent/pom.xml
+  download_gapic_generator_parent_pom "$gapic_generator_version"
+  grpc_version=$(grep grpc.version "parent-pom-$gapic_generator_version.xml" | sed 's/<grpc\.version>\(.*\)<\/grpc\.version>/\1/' | sed 's/^[[:space:]]*//;s/[[:space:]]*$//')
+  echo "$grpc_version"
+}
+
+get_protobuf_version() {
+  gapic_generator_version=$1
+  # get protobuf version from gapic-generator-java-pom-parent/pom.xml
+  download_gapic_generator_parent_pom "$gapic_generator_version"
+  protobuf_version=$(grep protobuf.version "parent-pom-$gapic_generator_version.xml" | sed 's/<protobuf\.version>\(.*\)<\/protobuf\.version>/\1/' | sed 's/^[[:space:]]*//;s/[[:space:]]*$//' | cut -d "." -f2-)
+  echo "$protobuf_version"
+}
+
+download_tools() {
+  gapic_generator_version=$1
+  protobuf_version=$2
+  grpc_version=$3
+  download_generator "$gapic_generator_version"
+  download_protobuf "$protobuf_version"
+  download_grpc_plugin "$grpc_version"
+}
+
+download_generator() {
+  gapic_generator_version=$1
+  cd "$working_directory"
+  if [[ "$gapic_generator_version" == *"-SNAPSHOT" ]]; then
+    # get SNAPSHOT from maven local repository.
+    generator_jar=$HOME/.m2/repository/com/google/api/gapic-generator-java/$gapic_generator_version/gapic-generator-java-$gapic_generator_version.jar
+    if [ ! -f "$generator_jar" ]; then
+      echo "Can't copy gapic-generator-java-$gapic_generator_version.jar from maven local repository."
+      exit 1
+    fi
+    cp "$generator_jar" "gapic-generator-java-$gapic_generator_version.jar"
+    echo "Copy gapic-generator-java-$gapic_generator_version.jar from maven local repository."
+  fi
+
+  if [ ! -f "gapic-generator-java-$gapic_generator_version.jar" ]; then
+    curl -LJ -o "gapic-generator-java-$gapic_generator_version.jar" https://repo1.maven.org/maven2/com/google/api/gapic-generator-java/"$gapic_generator_version"/gapic-generator-java-"$gapic_generator_version".jar
```

Comment:
Can we have proper error handling for all the downloading script? Either from Maven central or Github, we have a lot of scenarios in the past that these two website could be not responsive. In such case, we need to emit proper error message and exit the script.
Separately, we may want to consider downloading from [Google's Maven mirror](https://github.com/googleapis/sdk-platform-java/blob/main/settings.xml#L28) which could be faster and more consistent per @suztomo

Reply:
It does make the script simpler but I'm not sure executing `mvn` command is a good idea in this script. This adds one more pre-condition of running the script, and this also means we need to install Maven in anywhere we want to execute this script. Especially for internal presubmit checks, I'm not sure we have Maven installed.  
WDYT @meltsufin?

Reply:
If we don't use `mvn`, how can we download artifacts from [maven mirror repository](https://maven-central.storage-download.googleapis.com/maven2/)? It seems that the url pattern is not the same as maven central.

I can add steps to install `mvn` in the script if it's not in the environment.

Reply:
> It seems that the url pattern is not the same as maven central.

It's the same; otherwise Maven users cannot use the mirror.

Reply:
The url to download `com.google.api:gapic-generator-java:2.24.0` is `https://repo1.maven.org/maven2/com/google/api/gapic-generator-java/2.24.0/`.

However, `https://maven-central.storage-download.googleapis.com/maven2/com/google/api/gapic-generator-java/2.24.0/` doesn't return the right page. Is it because I don't have the right permission?

Reply:
> we need to emit proper error message and exit the script


@JoeWang1127 This is the concern Blake has. It's about the script may continue when the download fails. What's your answer?

Reply:
I modified the download command so the script can exit with an error message should the download fail.

Reply:
> The url to download com.google.api:gapic-generator-java:2.24.0 is https://repo1.maven.org/maven2/com/google/api/gapic-generator-java/2.24.0/.

This is not the JAR file.

Reply:
> > The url to download com.google.api:gapic-generator-java:2.24.0 is https://repo1.maven.org/maven2/com/google/api/gapic-generator-java/2.24.0/.
> 
> This is not the JAR file.

I changed the url to `https://maven-central.storage-download.googleapis.com/maven2/com/google/api/gapic-generator-java/2.24.0/gapic-generator-java-2.24.0.jar` now I can download the jar file. Thanks Tomo.

Then it comes back to whether we want to download dependencies using `mvn`.

Reply:
The concern that the curl command might continue when the download fails. Or the curl command appropriately fails with an error message.

How can you address the concern?


(Maven central or Google mirror is a trivial choice. Either one is fine)

Reply:
I changed the curl command to
```
curl -LJ -o "protobuf-$protobuf_version.zip" https://github.com/protocolbuffers/protobuf/releases/download/v"$protobuf_version"/protoc-"$protobuf_version"-linux-x86_64.zip || \
  download_fail "protobuf-$protobuf_version"
```

If the download fails, `download_fail` will print error message and exit the script.

Reply:
@suztomo I added `--fail` to curl command so the command return with non-zero exit code when the url is wrong or no file exists in remote server. Thanks for the help to clarify the question.

Regards the `mvn` command, I think it's simple to download artifacts from maven central (it will switch to maven local when downloading a SNAPSHOT version). I also tested that the command will return 1 if the version is wrong or the file doesn't exist. But I can change back to `curl` to remove `mvn` as a dependency. @blakeli0 @meltsufin WDYT?

Reply:
@JoeWang1127 In terms of hermetic build, can you make a (small) argument which is better between mvn and curl? Which one is more likely to fail (due to outdated configurations) after 2 years.

Reply:
Maven is useful in a java project, but we need to maintain a maven wrapper so that the generation is hermetic, this seems a overkill. 

On the other hand, `curl` is more flexible in shell project, so I decide to use curl to download dependencies.

Reply:
I made the change to use `curl --fail -m 30 --retry 2` to download the dependencies. The function will fail if no file exists or takes long time (> 30s) to finish. Also, use Google maven central mirror instead of maven central.

I'll resolve this comment and please reopen it if necessary.